### PR TITLE
Rust: Visibility check for qualified path resolution

### PR DESCRIPTION
**Pull request by hvitved**

### CODE REVIEW

File: rust/ql/lib/codeql/rust/internal/PathResolution.qll [hunk 1]
```diff
@@ -229,104 +234,125 @@ abstract class ItemNode extends Locatable {
...
     // items made available through `use` are available to nodes that contain the `use`
-    exists(UseItemNode use |
-      use = this.getASuccessor(_, _) and
-      result = use.getASuccessor(name, kind)
-    )
+    useOpt.asSome() =
+      any(UseItemNode use_ |
+        use_ = this.getASuccessor(_, _, _) and
+        result = use_.getASuccessor(name, kind, _)
+      )
```
Comment (geoffw0) [inline rust/ql/lib/codeql/rust/internal/PathResolution.qll:273]: I'm a bit nervous of these cases where we potentially have more than one `useOpt` around (three in this case, the value of the `any(...)` and the two `getASuccessor(, , _)`).  Can we imagine code where more than one of them has a value?

File: rust/ql/test/library-tests/path-resolution/my2/my3/mod.rs [hunk 2]
```diff
@@ -11,4 +11,4 @@ use super::nested6_f; // $ item=I116
 
 use super::*; // $ item=mod.rs
 
-trait MyTrait: Deref {} // $ MISSING: item=Deref
+trait MyTrait: Deref {} // $ item=Deref
```
Comment (geoffw0) [inline rust/ql/test/library-tests/path-resolution/my2/my3/mod.rs:14]: I'm a bit nervous that we appear to have only one (explicit) test case for this fairly complicated change.  Are there any edge cases we ought to be testing as well?

**Review by geoffw0 — COMMENTED** [review]
Good improvement on DCA.  :)

File: rust/ql/lib/codeql/rust/internal/PathResolution.qll (hunk 1 above)
Comment (hvitved) [reply rust/ql/lib/codeql/rust/internal/PathResolution.qll:273]: `useOpt` is meant to record the "nearest" `use` that make the item available, so in this case we can discard whatever values are in the two `getASuccessor` calls.

File: rust/ql/test/library-tests/path-resolution/my2/my3/mod.rs (hunk 2 above)
Comment (hvitved) [reply rust/ql/test/library-tests/path-resolution/my2/my3/mod.rs:14]: We also have some tests from https://github.com/github/codeql/pull/20453, but they didn't cover the case where an item from another crate was made available via a private `use` statement, which  the test case above does.

File: rust/ql/lib/codeql/rust/internal/PathResolution.qll (hunk 1 above)
Comment (geoffw0) [reply rust/ql/lib/codeql/rust/internal/PathResolution.qll:273]: Thank you for clarifying. :+1:

**Review by geoffw0 — APPROVED** [review]
I'm happy with this.

@paldepind might be a good idea for you to have a quick look over the changes as well at some point.
